src/codegen/generators/data_type_generator.rs

```diff
@@ -105,135 +21,153 @@ pub fn generate_data_type_stubs<'a>(llvm: &LLVM<'a>, index: &mut Index<'a>, data
+        }
+    }
+    for (name, user_type) in types {
+        let gen_type =  create_type(llvm, index, &types_index, name, user_type)?;
+        types_index.associate_type(name, gen_type)?
+    }
+    for (name, user_type) in types {
+        expand_opaque_types(llvm, index, &mut types_index, user_type)?;
+        if let Some(initial_value) = generate_initial_value(index, &types_index, llvm, user_type) {
+            types_index.associate_initial_value(name, initial_value)?
+        }
+    }
+    Ok(types_index)
+}
 
-                    //associate the enum element's global variable with the enume-element's name
-                    index.associate_global_variable(element, element_variable.as_pointer_value());
-                }
-            }
-            DataType::SubRangeType {
-                name,
-                referenced_type,
-            } => {
-                let alias_name = name.as_ref().map(|it| it.as_str()).unwrap();
-                index.associate_type_alias(alias_name, referenced_type.as_str());
-                if let Some(initializer) = &data_type.initializer {
-                    let generator = ExpressionCodeGenerator::new_context_free(llvm, index, None);
-                    let (_, initial_value) = generator.generate_expression(initializer)?;
-                    index.associate_type_initial_value(alias_name, initial_value);
-                }
+fn create_opaque_struct<'a>(llvm : &LLVM<'a>, data_type : &DataType) -> Option<BasicTypeEnum<'a>> {
+    if let DataTypeInformation::Struct {name, ..} = data_type.get_type_information() {
+        Some(llvm.create_struct_stub(name).into())
+    } else {
+        None
+    }
+}
+
+
+/// generates the members of an opaque struct and associates its initial values 
+fn expand_opaque_types<'ink>(llvm: &LLVM<'ink>, index : &Index, types_index : &mut LLVMTypedIndex<'ink>, data_type : &DataType) -> Result<(), CompileError> {
+    let information = data_type.get_type_information();
+    if let DataTypeInformation::Struct {member_names, .. } = information {
+            let mut struct_generator = StructGenerator::new(llvm, index, types_index);
+            let members = member_names.iter().map(|variable_name| index.find_member(data_type.get_name(), variable_name)
+            .unwrap())
+            .filter(|var| !var.is_return())
+            .collect();
+            let ((_, initial_value), member_values) = struct_generator.generate_struct_type(&members, data_type.get_name())?;
+            for (member, value) in member_values {
+                let qualified_name = format!("{}.{}", data_type.get_name(), member);
+                types_index.associate_initial_value(&qualified_name, value)?;
             }
-            DataType::ArrayType { name, .. } => {
-                generate_array_data_type(data_type, name, index, llvm, 
+            types_index.associate_initial_value(data_type.get_name(), initial_value)?;
+    }
+    Ok(())
+}
+
+//TODO : Recursion here might hurt us later
```

Comment:
I'd rather not keep this todo here

Reply:
Done. The idea was since we eagerly generate the types, it might have some performance effects but let's keep find this out when we do look at performance issues.